twister: add --check to validate test metadata

nashif opened the pull request:
for example, running this 
```./scripts/twister -T samples/ --check```

will result in:
```
QA      - filter with no integration platforms in samples/posix/gettimeofday/sample.posix.gettimeofday
QA      - skip set in samples/tfm_integration/psa_crypto/sample.psa_crypto
QA      - Too many excluded platforms in samples/userspace/shared_mem/sample.kernel.memory_protection.shared_mem: 6
QA      - platform_allow in samples/kernel/bootargs/sample.kernel.bootargs.multiboot without integration_platforms
QA      - build_only set in samples/sensor/mpu6050/sample.sensor.mpu6050
QA      - build_only set in samples/sensor/6dof_motion_drdy/sample.sensor.6dof_motion_drdy
QA      - build_only set in samples/sensor/mhz19b/sample.sensor.mhz19b
QA      - filter with no integration platforms in samples/sensor/mhz19b/sample.sensor.mhz19b
QA      - platform_allow in samples/sensor/clock/sample.sensor.clock.counter without integration_platforms
QA      - platform_allow in samples/sensor/clock/sample.sensor.clock.rtc without integration_platforms
QA      - filter with no integration platforms in samples/sensor/clock/sample.sensor.clock.system
QA      - build_only set in samples/sensor/max17262/sample.sensor.max17262
QA      - build_only set in samples/sensor/vl53l0x/sample.sensor.vl53l0x
QA      - filter with no integration platforms in samples/sensor/bmg160/sample.sensor.bmg160
QA      - Too many excluded platforms in samples/sensor/accel_trig/sample.sensor.accel_trig: 7
QA      - build_only set in samples/sensor/mcux_lpcmp/sample.sensor.mcux_lpcmp.no_trigger
QA      - build_only set in samples/sensor/icm42605/sample.sensor.icm42605
QA      - filter with no integration platforms in samples/sensor/tmp108/sample.sensor.tmp108
QA      - build_only set in samples/sensor/sht3xd/sample.sensor.sht3xd
QA      - build_only set in samples/sensor/sht3xd/sample.sensor.sht3xd.trigger
QA      - filter with no integration platforms in samples/sensor/max44009/sample.sensor.max44009
QA      - build_only set in samples/sensor/tdk_apex/sample.sensor.icm42670.apex
QA      - filter with no integration platforms in samples/sensor/lps22hh/sample.sensor.lps22hh
QA      - build_only set in samples/sensor/ms5837/sample.sensor.ms5837
QA      - filter with no integration platforms in samples/sensor/th02/sample.sensor.th02
QA      - Too many integration platforms in samples/sensor/accel_polling/sample.sensor.accel_polling: 15
QA      - filter with no integration platforms in samples/sensor/tmp112/sample.sensor.tmp112
QA      - filter with no integration platforms in samples/sensor/co2_polling/sample.sensor.co2
QA      - Unsupported depends_on {'uart_interrupt_driven', 'serial'} in samples/sensor/co2_polling/sample.sensor.co2
QA      - filter with no integration platforms in samples/sensor/max6675/sample.sensor.max6675
QA      - build_only set in samples/sensor/ina219/sample.drivers.ina219
QA      - filter with no integration platforms in samples/sensor/sx9500/sample.sensor.sx9500
QA      - filter with no integration platforms in samples/sensor/stream_fifo/sample.sensor.stream_fifo
QA      - build_only set in samples/sensor/lsm6dso_i2c_on_i3c/sample.sensor.lsm6dso.i2c_on_i3c_bus
QA      - filter with no integration platforms in samples/sensor/lsm6dso_i2c_on_i3c/sample.sensor.lsm6dso.i2c_on_i3c_bus
QA      - build_only set in samples/sensor/grow_r502a/sample.sensor.grow_r502a
QA      - filter with no integration platforms in samples/sensor/grow_r502a/sample.sensor.grow_r502a
QA      - Unsupported depends_on {'lis2dh', 'i2c'} in samples/sensor/lis2dh/sample.sensor.lis2dh
QA      - filter with no integration platforms in samples/sensor/veaa_x_3/sample.sensor.veaa_x_3
QA      - build_only set in samples/sensor/sgp40_sht4x/sample.sensor.sgp40_sht4x
QA      - build_only set in samples/sensor/mcux_acmp/sample.sensor.mcux_acmp.no_trigger
QA      - build_only set in samples/sensor/amg88xx/sample.sensor.amg88xx.amg88xx_grid_eye_eval_shield
QA      - build_only set in samples/sensor/amg88xx/sample.sensor.amg88xx.amg88xx_eval_kit
QA      - filter with no integration platforms in samples/sensor/qdec/sample.sensor.qdec_sensor
QA      - platform_allow in samples/sensor/qdec/sample.sensor.sam_qdec_sensor without integration_platforms
QA      - Unsupported depends_on {'spi', 'bme280'} in samples/sensor/bme280/sample.sensor.bme280.spi
QA      - filter with no integration platforms in samples/sensor/soc_voltage/sample.sensor.soc_voltage
QA      - build_only set in samples/sensor/lps22hh_i3c/sample.sensor.lps22hh.i3c
QA      - filter with no integration platforms in samples/sensor/lps22hh_i3c/sample.sensor.lps22hh.i3c
```



Note: many of the limits and guards are arbitrary, they will be configurable somewhere.

nashif committed:
twister: move harnesses into constants and check tests

Check if harness used in test metadata is something we support and issue
and qa warning when an unknown harness is being used. If harness is not
supported, this means the test will only build and we have no way to run
it in CI.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

PerMac commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -549,6 +554,46 @@ def add_testsuites(self, testsuite_filter=None):
+                                logger.qa(f"platform_allow in {suite.name} without integration_platforms")
+
+
+                            if suite.build_only and not "build" in suite.id:
```
some more comments would help to understand the guidelines. Is the idea behind it that all scenarios with "build_only: true" have "build" in their name? If so, the printed output doesn't reflect this.

nashif replied:
that is a hack right now, there are build tests, i.e. things we only build and should NEVER run and we need to mark those differently and distinguish them from build_only set for other reasons or no reason at all. I am thinking to introduce and enforce 'harness: build` for such tests and stop using `build_only`...

PerMac commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -549,6 +554,46 @@ def add_testsuites(self, testsuite_filter=None):
+                                logger.qa(f"Unsupported harness {suite.harness} in {suite.name}")
+
+                            if suite.filter and not suite.integration_platforms and not suite.platform_allow:
+                                logger.qa(f"filter with no integration platforms in {suite.name}")
```
the condition also depends on platform_allow. Why? The printed output doesn't match this.

PerMac commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -549,6 +554,46 @@ def add_testsuites(self, testsuite_filter=None):
+                            if suite.filter and not suite.integration_platforms and not suite.platform_allow:
+                                logger.qa(f"filter with no integration platforms in {suite.name}")
+
+                            if suite.platform_allow and suite.integration_platforms:
+                                _default_p = set(self.default_platforms)
+                                _platform_allow = set(suite.platform_allow)
+                                _integration_p = set(suite.integration_platforms)
+                                _intersection1 = _default_p.intersection(_platform_allow)
+                                _intersection2 = _default_p.intersection(_integration_p)
+
+                                if  _intersection1 and not _intersection2:
+                                    logger.qa(f"No default platforms in integration_platform for {suite.name}")
```
again, looks like some more complex condition, where the output doesn't match exactly the logic in code. This triggers only when there is platform_allow list and it includes at list one platform from default_platforms and there is also integration_platforms, but doesn't include any from default_platforms. Correct? What is the reasoning behind it? Encouraging usage of default_platforms in integration_platforms?

PerMac commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -549,6 +554,46 @@ def add_testsuites(self, testsuite_filter=None):
+                                if  _intersection1 and not _intersection2:
+                                    logger.qa(f"No default platforms in integration_platform for {suite.name}")
+
+                            _do = set(suite.depends_on).intersection(self.supported_features)
+                            if len(_do) != len(suite.depends_on):
+                                logger.qa(f"Unsupported depends_on {suite.depends_on} in {suite.name}")
```
IMO we would have better control and understanding if we extract all existing supported_features and depends_on in advance and add them to constants module instead of recreating the list of supported_features each time. Then we will always see when something new/wrong is being added.

The logic could work both way: having depends_on which no platform supports, but also platform having a supported key which is not used in any test as depends_on

nashif replied:
if we want to put this in a constant is open for discussion, right now this tries to use the data as we have it.

PerMac commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -578,6 +623,9 @@ def add_testsuites(self, testsuite_filter=None):
                         else:
                             self.testsuites[suite.name] = suite
 
+                    if scenario_count > 5 and self.options.log_level == "QA":
```
`and self.options.log_level == "QA"` looks redundant,

nashif replied:
those will be cleaned up, I was getting the wrong level at some point and did not want the messages to interfere.

PerMac commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -549,6 +554,46 @@ def add_testsuites(self, testsuite_filter=None):
                                 suite.platform_allow,
                                 f"platform_allow in {suite.name}")
 
+                        if self.options.log_level == "QA":
```
this `if` seems redundant. Using logger.qa already handles the logic when to print

nashif replied:
ditto

PerMac reviewed (commented):
I think it would be easier to maintain and have a better understanding of what is done in the code when all the QA functionality is moved to it's own module and its functions are called in limited (single?) places.

nashif commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -549,6 +554,46 @@ def add_testsuites(self, testsuite_filter=None):
                                 f"platform_allow in {suite.name}")
 
+                        if self.options.log_level == "QA":
+                            if len(suite.platform_allow) > 10:
```
read the note at the end:

```Note: many of the limits and guards are arbitrary, they will be configurable somewhere.```

golowanow replied:
it can be already a directory with constants at one place somewhere instead of these magic values.

nashif replied:
done in  the test configuration file.

nashif commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -801,7 +801,7 @@ def add_parse_arguments(parser = None) -> argparse.ArgumentParser:
         type=str.upper,
         default='INFO',
-        choices=['NOTSET', 'DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'],
+        choices=['NOTSET', 'DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL', 'STEP', 'QA'],
```
> Do we need the 'STEP' level? Maybe such messages can be part of DEBUG?

DEBUG is very busy and the STEP is for debugging the queue management and focuses on that. This can also be printed in debug mode, but isolation is needed for debugging twister core functionality, not necessarily to debug test execution.

golowanow replied:
> To me this QA messaging looks like an independent feature, that could be turned on/off regardless of log-level set.

I also think so.

nashif commented on scripts/pylib/twister/twisterlib/constants.py:
```diff
@@ -16,5 +16,21 @@
+]
+
+SUPPORTED_HARNESS = SUPPORTED_RUNNABLE_HARNESS + SUPPORTED_BUILDABLE_HARNESS
+
```
Sure, but not part of this change, all of this can be added later.

nashif commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -444,10 +446,12 @@ def add_configurations(self):
         for platform in generate_platforms(board_roots, soc_roots, arch_roots):
             if not platform.twister:
                 continue
+            self.supported_features.extend(platform.supported)
```
this is supposed to be a set

nashif commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -549,6 +554,46 @@ def add_testsuites(self, testsuite_filter=None):
+                                logger.qa(f"Too many excluded platforms in {suite.name}: {len(suite.platform_exclude)}")
+
+                            if len(suite.platform_allow) > 1 and not suite.integration_platforms:
+                                logger.qa(f"platform_allow in {suite.name} without integration_platforms")
```
if you have 1 entry in the platform_allow it does not make sense to replicate it into integration_platform, afaik it will have the same effect.

PerMac replied:
"afaik it will have the same effect." there will be some small differences. E.g. when you use integration_platforms a given configuration cannot get skipped (because of one of dozens cases in twister) in most cases as it will be changed to error. Without it there is no such warranty, so you can end up with a configuration that cannot be verified. 

My question can be reworded: "why we need integration_platform when we have platform_allow, and don't need it when there is no platform_allow (unless there is filter)".
Let's decide in which cases integration_platforms are required (if not in all, which would simplify the logic) and then we can adjust the logic to match it.

golowanow commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -801,7 +801,7 @@ def add_parse_arguments(parser = None) -> argparse.ArgumentParser:
         type=str.upper,
         default='INFO',
-        choices=['NOTSET', 'DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'],
+        choices=['NOTSET', 'DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL', 'STEP', 'QA'],
```
```suggestion
        choices=['NOTSET', 'DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL', 'TRACE', 'SANITY'],
```
imo these are better reflect the purpose.

golowanow commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -985,6 +985,7 @@ def process(self, pipeline, done, message, lock, results):
         self.instance.setup_handler(self.env)
 
         if op == "filter":
+            logger.step(f"Applying filters {self.instance.name}")
```
in all these places, just the same
```suggestion
            logger.step(f"op:{op} instance:{self.instance.name}")
```
this allows to grep/parse/filter logs much easier than a bunch of new and different entries.

golowanow commented on tests/kernel/fatal/no-multithreading/testcase.yaml:
```diff
@@ -14,8 +14,7 @@ common:
   integration_platforms:
-    - qemu_cortex_m3
-    - nsim/nsim_em
+    - qemu_riscv32
```
is it a leftover from some other change ? it looks not related to this commit/PR

nashif commented:
Another batch of fixes based on the results from this PR: #86716

nashif committed:
twister: add --check to validate test metadata

Add new --check option to verify tests for quality of meta-data and
correctness.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
so, to shift focus from the messanger to the message, I removed the logging level thing and added some summary and custom handling:

now, get the results with `./scripts/twister --check`, output:

```
WARNING - skip: skip set in tests/arch/arm64/arm64_gicv3_its/arch.arm64.gicv3_its
WARNING - skip: skip set in tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform
WARNING - skip: skip set in tests/lib/cmsis_dsp/filtering/libraries.cmsis_dsp.filtering
WARNING - skip: skip set in tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix
WARNING - skip: skip set in tests/subsys/logging/log_stack/logging.stack.deferred
WARNING - skip: skip set in tests/subsys/logging/log_stack/logging.stack.deferred_no_opt
WARNING - skip: skip set in tests/subsys/logging/log_stack/logging.stack.immediate
WARNING - skip: skip set in tests/subsys/logging/log_stack/logging.stack.immediate_no_opt
WARNING - skip: skip set in tests/subsys/logging/log_stack/logging.stack.deferred_cpp
WARNING - skip: skip set in tests/subsys/logging/log_stack/logging.stack.deferred_cpp_no_opt
WARNING - skip: skip set in tests/subsys/logging/log_stack/logging.stack.immediate_cpp
WARNING - skip: skip set in tests/subsys/logging/log_stack/logging.stack.immediate_cpp_no_opt
WARNING - skip: skip set in samples/tfm_integration/psa_crypto/sample.psa_crypto
WARNING - skip: skip set in samples/drivers/ipm/ipm_esp32/sample.ipm.ipm_esp32
WARNING - skip: skip set in samples/subsys/shell/fs/sample.filesystem.shell.fuse
ERROR   - unsupported_depends_on: Unsupported depends_on {'pinctrl', 'nrf'} in tests/drivers/pinctrl/nrf/drivers.pinctrl.nrf
ERROR   - unsupported_depends_on: Unsupported depends_on {'zms'} in tests/subsys/settings/zms/settings.zms
ERROR   - unsupported_depends_on: Unsupported depends_on {'serial', 'uart_interrupt_driven'} in samples/sensor/co2_polling/sample.sensor.co2
ERROR   - unsupported_depends_on: Unsupported depends_on {'i2c', 'lis2dh'} in samples/sensor/lis2dh/sample.sensor.lis2dh
ERROR   - unsupported_depends_on: Unsupported depends_on {'spi', 'bme280'} in samples/sensor/bme280/sample.sensor.bme280.spi
ERROR   - unsupported_depends_on: Unsupported depends_on {'mailbox'} in samples/boards/st/h7_dual_core/sample.boards.stm32.h7_dual_core
ERROR   - unsupported_depends_on: Unsupported depends_on {'gnss'} in samples/drivers/gnss/sample.drivers.gnss
ERROR   - unsupported_depends_on: Unsupported depends_on {'charger'} in samples/drivers/charger/sample.drivers.charger
INFO    - Summary of checks:
INFO    -
+------------------------+------------+---------------+----------------------------------------------+
| Check                  | Severity   |   Occurrences | Description                                  |
+========================+============+===============+==============================================+
| harness                | warning    |           451 | Unsupported harness                          |
+------------------------+------------+---------------+----------------------------------------------+
| build_only             | warning    |           359 | build_only set                               |
+------------------------+------------+---------------+----------------------------------------------+
| filter_no_integration  | error      |           277 | filter with no integration platforms         |
+------------------------+------------+---------------+----------------------------------------------+
| allow_no_integration   | error      |           223 | allow_platform without integration_platforms |
+------------------------+------------+---------------+----------------------------------------------+
| no_default_integration | error      |           140 | No default platforms in integration_platform |
+------------------------+------------+---------------+----------------------------------------------+
| scenario_count         | warning    |           123 | Too many scenarios                           |
+------------------------+------------+---------------+----------------------------------------------+
| allow_platform         | warning    |            42 | Too many entries in allow_platform           |
+------------------------+------------+---------------+----------------------------------------------+
| exclude_platform       | warning    |            36 | Too many excluded platforms                  |
+------------------------+------------+---------------+----------------------------------------------+
| integration_platform   | warning    |            21 | Too many integration platforms               |
+------------------------+------------+---------------+----------------------------------------------+
| skip                   | warning    |            15 | skip set                                     |
+------------------------+------------+---------------+----------------------------------------------+
| unsupported_depends_on | error      |             8 | Unsupported depends_on                       |
+------------------------+------------+---------------+----------------------------------------------+
```

hakehuang commented:
May I raise a request here? can I check the specific platform only? For now it seems to give an overall result.

nashif commented:
> May I raise a request here? can I check the specific platform only? For now it seems to give an overall result.

not sure what you mean

PerMac commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -78,6 +81,92 @@ class TestLevel:
     levels = []
     scenarios = []
 
+class Checker:
```
personal opinion (not a request for change): I find this class and how it is handled hard to follow and it took me some time to understand what is going on. Especially when the tuples defined here are added to a `checks` list every time a given type of issue is found. I believe that using named fields would be easier to understand than using a logic based on a positional argument within a nested structure.

PerMac commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -548,6 +646,72 @@ def add_testsuites(self, testsuite_filter=None):
+                                    f"{len(suite.platform_exclude)}"
+                                )
+
+                            if len(suite.platform_allow) > 1 and not suite.integration_platforms:
+                                checker.add(
+                                    checker.ALLOW_NO_INTEGRATION,
+                                    f"platform_allow in {suite.name} without integration_platforms"
+                                )
```
not sure how helpful is this check. Why not just checking if integration_platforms are provided or not (regardless of platform_allow)? Or from another perspective: when it is not an issue when there are no integration_platforms?

PerMac commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -548,6 +646,72 @@ def add_testsuites(self, testsuite_filter=None):
+                                    checker.HARNESS,
+                                    f"Unsupported harness {suite.harness} in {suite.name}"
+                                )
+
+                            if (suite.filter and not suite.integration_platforms and
+                                not suite.platform_allow and not suite.platform_key):
+                                checker.add(
+                                    checker.FILTER_NO_INTEGRATION,
+                                    f"filter with no integration platforms in {suite.name}"
```
the reason string doesn't correspond to the logic (not mentioning platform_key and platform_allow)

PerMac reviewed (commented):
In general looks useful. It is hard to fully evaluate the check when it is not clear what are the QA guidelines. Especially when the check's logic is more complex and doesn't match printed output.

hakehuang commented:
> not sure what you mean
@nashif  if I run
`./scripts/twister -p frdm_k64f -T samples/ --check`
it will only give me the frdm_k64f related check
